File: QytetetJava/src/modeloqytetet/Calle.java
/**
 * @author Elena Cantero Molina
 */
package modeloqytetet;

public class Calle extends Casilla{
    private int numHoteles;
    private int numCasas;
    TituloPropiedad titulo = null;
    
    Calle(int numeroCasilla, int coste, TituloPropiedad titulo){
        super(numeroCasilla, coste, TipoCasilla.CALLE);
        numHoteles = 0;
        numCasas = 0;
        SetTituloPropiedad(titulo);
        this.titulo.casilla = this;
    }
    
    public TituloPropiedad getTitulo(){ return titulo; }
    
    public int GetNumeroCasas(){ return this.numCasas; }

    public int GetNumeroHoteles(){ return this.numHoteles; }
    
    public TituloPropiedad GetTitulo(){ return this.titulo; }
    
    public void SetNumCasas(int numCasas){ this.numCasas = numCasas; }

    public void SetNumHoteles(int numHoteles){ this.numHoteles = numHoteles; }

    private void SetTituloPropiedad(TituloPropiedad titulo){ this.titulo = titulo; }
    
    TituloPropiedad asignarPropietario(Jugador jugador){
        titulo.setPropietario(jugador);
        titulo.propietario = jugador;
        return titulo;
    }
    
    int calcularValorHipoteca(){ return titulo.GetHipotecaBase()  + (int)(numCasas*0.5 * titulo.GetHipotecaBase() + numHoteles * titulo.GetHipotecaBase()); }

    int cancelarHipoteca(){
        titulo.SetHipotecada(false);
        return ((int) (1.1 * calcularValorHipoteca()));
    }

    int cobrarAlquiler(){
        int costeAlquilerBase = titulo.GetAlquilerBase();
        titulo.cobrarAlquiler(costeAlquilerBase + (int)(numCasas*0.5 + numHoteles*2));
        return costeAlquilerBase;
    }    

    int edificarCasa(){
        this.SetNumCasas(numCasas+1);
        return titulo.GetPrecioEdificar();
    }

    int edificarHotel(){
        this.SetNumHoteles(numHoteles+1);
        return titulo.GetPrecioEdificar();
    }    
    
    boolean estaHipotecada(){ return this.titulo.GetHipotecada(); }
    
    int getPrecioEdificar(){ return titulo.GetPrecioEdificar(); }    
    
    int hipotecar(){ 
        titulo.SetHipotecada(true);
        return this.calcularValorHipoteca(); 
    }
    
    int precioTotalComprar(){
        int precio = GetCoste() + (numCasas + numHoteles)*titulo.GetPrecioEdificar();
        precio = precio + (int) (precio*titulo.GetFactorRevalorizacion());
        return precio;
    }    
    
    boolean propietarioEncarcelado(){ return this.titulo.propietarioEncarcelado(); }

    public boolean sePuedeEdificarCasa(){        
        if (Jugador.factorEspeculador == 2){
            return numCasas < 8;
        }
        else{
            return numCasas < 4;
        }
    }

    boolean sePuedeEdificarHotel(){
        if (Jugador.factorEspeculador == 2){
            return numCasas==8 && numHoteles < 8;
        }
        else{
            return numCasas==4 && numHoteles < 4;
        }
    }  
    
    boolean tengoPropietario(){ return this.titulo.tengoPropietario(); }    

    int venderTitulo(){
        int precioComprar = GetCoste() + (numCasas + numHoteles)*titulo.GetPrecioEdificar();
        titulo.setPropietario(null);
        this.SetNumCasas(0);
        this.SetNumHoteles(0);
        return precioComprar + (int)(titulo.GetFactorRevalorizacion()*precioComprar);
    }
    
    @Override
    public String toString() {
        return "Calle -> " + "\ntitulo -> " + this.GetTitulo()+ "\nnumHoteles -> " + this.GetNumeroHoteles() + "\nnumCasas -> " + this.GetNumeroCasas() + '\n';
    }

}

File: QytetetRuby/lib/calle.rb
#Elena Cantero Molina
#encoding: utf-8
#trig.rb

module ModeloQytetet
  class Calle < Casilla
    attr_accessor :numero_hoteles, :numero_casas, :titulo
    
    def initialize(numero_casilla, coste, titulo)
      super(numero_casilla, coste, TipoCasilla::CALLE)
      @numero_hoteles = 0
      @numero_casas = 0
      @titulo = titulo
      
      if !@titulo.nil?
        @titulo.casilla = self
      end
    end
    
    def asignar_propietario(jugador)
      @titulo.propietario = jugador
    end 
    
    def calcular_valor_hipoteca
      return (@titulo.hipotecaBase + (@numero_casas*0.5*@titulo.hipotecaBase + @numero_hoteles*@titulo.hipotecaBase))
    end
    
    def cancelar_hipoteca
      @titulo.hipotecada = false
      return 1.1*calcular_valor_hipoteca
    end
    
    def cobrar_alquiler
      coste_alquiler_base = @titulo.alquilerBase
      @titulo.cobrar_alquiler(coste_alquiler_base + @numero_casas*0.5 + @numero_hoteles*2)
      return coste_alquiler_base
    end
    
    def edificar_casa
      @numero_casas = @numero_casas + 1
      return @titulo.precioEdificar
    end
    
    def edificar_hotel
      @numero_hoteles = @numero_hoteles + 1
      return @titulo.precioEdificar
    end
    
    def esta_hipotecada
      return @titulo.hipotecada.eql?(true)
    end
    
    def get_precio_edificar
      return @titulo.precioEdificar
    end
    
    def hipotecar
      @titulo.hipotecada = true
      return calcular_valor_hipoteca
    end
    
    def precio_total_comprar
        precio = @coste + (@numero_casas + @numero_hoteles)*@titulo.precioEdificar
        precio += precio*@titulo.factorRevalorizacion
        return precio;
    end
    
    def propietario_encarcelado
      return @titulo.propietario_encarcelado
    end
    
    def se_puede_edificar_casa(factor_especulador)
      if (factor_especulador == 1)
        return @numero_casas < 4
      else
        return @numero_casas < 8
      end
    end
    
    def se_puede_edificar_hotel(factor_especulador)
      if (factor_especulador == 1)
        return (@numero_hoteles < 4 and @numero_casas == 4)
      else
        return (@numero_hoteles < 8 and @numero_casas == 8)
      end
    end
    
    def tengo_propietario
      return @titulo.tengo_propietario
    end 
    
    def vender_titulo
      precio_comprar = @coste + (@numero_casas + @numero_hoteles)*@titulo.precioEdificar
      @propietario = nil
      @numero_casas = 0
      @numero_hoteles = 0
      return precio_comprar + (@titulo.factorRevalorizacion*precio_comprar)
    end
    
    def to_s
      "#{super.to_s}\nNumero Casas -> #{@num_casas}\nNumero Hoteles -> #{@num_hoteles}\nTitulo Propiedad -> \n#{@titulo}"
    end
  end
end
